### PR TITLE
Cover non-empty-array in `array_merge`

### PR DESCRIPTION
**Pull request by staabm**
Refs https://github.com/phpstan/phpstan/issues/5327

### CODE REVIEW

File: tests/PHPStan/Analyser/LegacyNodeScopeResolverTest.php [hunk 1]
```diff
@@ -5129,35 +5129,35 @@ public function dataArrayFunctions(): array
...
 			],
 			[
-				'array<int|non-empty-string, stdClass>',
+				'array<int|non-empty-string, stdClass>&nonEmpty',
 				'array_merge($stringOrIntegerKeys)',
 			],
 			[
-				'array<int|string, DateTimeImmutable|int>',
+				'array<int|string, DateTimeImmutable|int>&nonEmpty',
 				'array_merge($generalStringKeys, $generalDateTimeValues)',
```
Comment (staabm) [inline tests/PHPStan/Analyser/LegacyNodeScopeResolverTest.php:5137]: Im am not confident, whether this is a expected change, or the impl is buggy for this case.

File: tests/PHPStan/Analyser/LegacyNodeScopeResolverTest.php [hunk 2]
```diff
@@ -5129,35 +5129,35 @@ public function dataArrayFunctions(): array
...
 			],
 			[
-				'array<int|non-empty-string, 2|4|\'a\'|\'b\'|\'green\'|\'red\'|\'trapezoid\'>',
+				'array<int|non-empty-string, 2|4|\'a\'|\'b\'|\'green\'|\'red\'|\'trapezoid\'>&nonEmpty',
 				'array_merge(array("color" => "red", 2, 4), array("a", "b", "color" => "green", "shape" => "trapezoid", 4))',
 			],
 			[
-				'array<int|string, DateTimeImmutable|int>',
+				'array<int|string, DateTimeImmutable|int>&nonEmpty',
 				'array_merge(...[$generalStringKeys, $generalDateTimeValues])',
```
Comment (staabm) [inline tests/PHPStan/Analyser/LegacyNodeScopeResolverTest.php:5161]: Im am not confident, whether this is a expected change, or the impl is buggy for this case.

File: src/Type/Php/ArrayMergeFunctionDynamicReturnTypeExtension.php [hunk 3]
```diff
@@ -42,12 +44,24 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 
 			$keyTypes[] = TypeUtils::generalizeType($argType->getIterableKeyType(), GeneralizePrecision::moreSpecific());
 			$valueTypes[] = $argType->getIterableValueType();
+
+			if (!$argType->isIterableAtLeastOnce()->yes()) {
+				continue;
+			}
+
+			$nonEmpty = true;
```
Comment (staabm) [inline src/Type/Php/ArrayMergeFunctionDynamicReturnTypeExtension.php:52]: this doesn't make much sense to me in this case, but phpcbf enforced the early-continue construct

**Comment by staabm** [comment]
ready to merge

**Comment by ondrejmirtes** [comment]
Thank you.

**Pull request merged**